File: ui/src/main/java/org/wcl/mfn/ui/controller/tools/ContractCalculatorController.java
package org.wcl.mfn.ui.controller.tools;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.PropertySource;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.GetMapping;
import org.wcl.mfn.config.ui.common.PageTitleConfig;
import org.wcl.mfn.config.ui.mvc.ModelConfig;
import org.wcl.mfn.config.ui.mvc.ViewConfig;
import org.wcl.mfn.config.ui.tools.ContractCalculatorConfig;
import org.wcl.mfn.config.ui.tools.ContractCalculatorMessagesConfig;
import org.wcl.mfn.ui.model.tools.contractcalculator.SuggestedYearContract;

import java.util.List;

@Controller
@PropertySource("classpath:/url/url.properties")
public class ContractCalculatorController {

    @Autowired
    private PageTitleConfig pageTitleConfig;

    @Autowired
    private ViewConfig viewConfig;

    @Autowired
    private ModelConfig modelConfig;

    @Autowired
    private ContractCalculatorConfig contractCalculatorConfig;
    @Autowired
    private ContractCalculatorMessagesConfig contractCalculatorMessagesConfig;

    @GetMapping("#{'${url.tools.contract-calculator}'}")
    public String contractCalculatorPage(Model model) {
        model.addAttribute(modelConfig.pageTitle(), pageTitleConfig.contractCalculatorPageTitle());
        model.addAttribute(contractCalculatorConfig.contractParameters(), generateBlankContractParameters());
        model.addAttribute("contract_parameter_elements", generateContractCalculatorUIElements());
        model.addAttribute(contractCalculatorConfig.suggestedContracts(),generateBlankContracts());

        return viewConfig.contractCalculatorView();
    }

    private List<SuggestedYearContract> generateBlankContracts() {
       return List.of(new SuggestedYearContract(2,0,0,
                            List.of(0,0),0,
                            List.of(0,0)),
                      new SuggestedYearContract(3,0,0,
                            List.of(0,0, 0),0,
                              List.of(0,0,0)),
                      new SuggestedYearContract(4,0,0,
                            List.of(0,0, 0, 0),0,
                              List.of(0,0,0,0)),
                      new SuggestedYearContract(5,0,0,
                            List.of(0,0, 0, 0, 0),0,
                              List.of(0,0,0,0,0)),
                      new SuggestedYearContract(6,0,0,
                            List.of(0,0, 0, 0, 0, 0),0,
                              List.of(0,0,0,0,0,0)));
    }


    private ContractParameters generateBlankContractParameters() {
        return new ContractParameters(0,0);
    }

    private ContractCalculatorUIElements generateContractCalculatorUIElements() {
        return new ContractCalculatorUIElements(contractCalculatorMessagesConfig.minSalary(),
                                                contractCalculatorMessagesConfig.bonus(),
                new Escalator(contractCalculatorMessagesConfig.escalator(),
                                contractCalculatorConfig.escalatorDefault(),
                        contractCalculatorConfig.escalatorMax(),
                        contractCalculatorConfig.escalatorStep()
//                        ccc.

                        ),
        contractCalculatorMessagesConfig.reset(),
                contractCalculatorMessagesConfig.submit());
    }

    record ContractParameters(int salary, int bonus, int escalator) {
        private static final int DEFAULT_ESCALATOR = 5;

        public ContractParameters(int salary, int bonus) {
            this(salary, bonus, DEFAULT_ESCALATOR);
        }
    }

    record Escalator(String label, int defaultValue, int max, int step) {}

    record ContractCalculatorUIElements(String minSalary, String bonus, Escalator escalator, String reset, String submit) {
    }
}

File: ui/src/main/kotlin/org/wcl/mfn/ui/controller/tools/ContractCalculatorController.kt
package org.wcl.mfn.ui.controller.tools

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.PropertySource
import org.springframework.stereotype.Controller
import org.springframework.ui.Model
import org.springframework.web.bind.annotation.GetMapping
import org.wcl.mfn.config.ui.common.PageTitleConfig
import org.wcl.mfn.config.ui.mvc.*
import org.wcl.mfn.config.ui.tools.*
import org.wcl.mfn.ui.model.tools.contractcalculator.SuggestedYearContract

@Controller
@PropertySource("classpath:/url/url.properties")
class ContractCalculatorController {
    @Autowired
    private val pageTitleConfig: PageTitleConfig? = null

    @Autowired
    private val viewConfig: ViewConfig? = null

    @Autowired
    private val modelConfig: ModelConfig? = null

    @Autowired
    private val contractCalculatorConfig: ContractCalculatorConfig? = null

    @Autowired
    private val contractCalculatorMessagesConfig: ContractCalculatorMessagesConfig? = null

    @GetMapping("#{'\${url.tools.contract-calculator}'}")
    fun contractCalculatorPage(model: Model): String? {
        modelConfig!!.pageTitle()?.let { model.addAttribute(it, pageTitleConfig!!.contractCalculatorPageTitle()) }
        contractCalculatorConfig!!.contractParameters()
            ?.let { model.addAttribute(it, generateBlankContractParameters()) }
        model.addAttribute("contract_parameter_elements", generateContractCalculatorUIElements())
        contractCalculatorConfig.suggestedContracts()?.let { model.addAttribute(it, generateBlankContracts()) }

        return viewConfig!!.contractCalculatorView()
    }

    private fun generateBlankContracts(): List<SuggestedYearContract> {
        return java.util.List.of(
            SuggestedYearContract(
                2, 0, 0,
                listOf(0, 0), 0,
                listOf(0, 0)
            ),
            SuggestedYearContract(
                3, 0, 0,
                listOf(0, 0, 0), 0,
                listOf(0, 0, 0)
            ),
            SuggestedYearContract(
                4, 0, 0,
                listOf(0, 0, 0, 0), 0,
                listOf(0, 0, 0, 0)
            ),
            SuggestedYearContract(
                5, 0, 0,
                listOf(0, 0, 0, 0, 0), 0,
                listOf(0, 0, 0, 0, 0)
            ),
            SuggestedYearContract(
                6, 0, 0,
                listOf(0, 0, 0, 0, 0, 0), 0,
                listOf(0, 0, 0, 0, 0, 0)
            )
        )
    }


    private fun generateBlankContractParameters(): ContractParameters {
        return ContractParameters(0, 0)
    }

    private fun generateContractCalculatorUIElements(): ContractCalculatorUIElements {
        return ContractCalculatorUIElements(
            contractCalculatorMessagesConfig!!.minSalary(),
            contractCalculatorMessagesConfig.bonus(),
            Escalator(
                contractCalculatorMessagesConfig.escalator(),
                contractCalculatorConfig!!.escalatorDefault(),
                contractCalculatorConfig.escalatorMax(),
                contractCalculatorConfig.escalatorStep()
            ),
            contractCalculatorMessagesConfig.reset(),
            contractCalculatorMessagesConfig.submit()
        )
    }

  //  @JvmRecord
    internal data class ContractParameters(val salary: Int, val bonus: Int, val escalator: Int) {
        constructor(salary: Int, bonus: Int) : this(salary, bonus, DEFAULT_ESCALATOR)

        companion object {
            private const val DEFAULT_ESCALATOR = 5
        }
    }

   // @JvmRecord
    internal data class Escalator(val label: String?, val defaultValue: Int, val max: Int, val step: Int)

  //  @JvmRecord
    internal data class ContractCalculatorUIElements(
        val minSalary: String?,
        val bonus: String?,
        val escalator: Escalator,
        val reset: String?,
        val submit: String?
    )
}


